clis/tcli/src/bin/tclid/coms.rs

```diff
@@ -0,0 +1,152 @@
+
+                    if let Some(response) = self.listener_chan.rx.recv().await {
+                        match response {
+                            IPCMessage::Command(_) => {
+                                eprint!("The listener got a command from the event_loop which should never happen, so ignoring.");
+                            }
+                            IPCMessage::Response(response) => {
+                                stream.write_all(response.as_bytes()).await?
+                            }
+                            IPCMessage::QuitResponse(response) => {
+                                stream.write_all(response.as_bytes()).await?;
+                                break Ok(());
+                            }
+                        }
+                    }
+                }
+                _ => {
+                    eprintln!("Event loop had stopped responding to ipc requests, so stopping IPC listener");
+                    break Ok(());
+                }
+            }
+        }
+    }
+
+    /// Function for sending commands to the server. Utilized by the API so purposefully made non-async to avoid overhead.
+    /// Blocks while waiting for a response, but that's intended behavior, because the API does not use an async runtime.
+    /// Also retrieves the response from the IPC server.
+    /// Note that endlines are used to indicate the end of a message so the server knows when to stop reading and
+    /// send a response, otherwise the server would wait until the API closes the socket and won't be able to send a response.
+    /// This is also sort of a deadlock risk, so should be handled with care.
+    ///
+    /// # Arguments
+    ///
+    /// * `addr` - Address (more specifically path, but left address for clarity) of the IPC socket.
+    /// * `cmd` - Command to be sent to the daemon.
+    ///
+    /// # Returns
+    ///
+    /// A Result containing a response string.
+    pub fn send_command(addr: &Path, cmd: &str) -> Result<String> {
+        let mut response_buffer = String::new();
+        let mut stream = BlockingSocketStream::connect(addr)?;
+        writeln!(stream, "{}", cmd)?;
+        stream.read_to_string(&mut response_buffer)?;
+
+        Ok(response_buffer)
+    }
+
+    /// Function Implementing custom retry logic for send_command, because the public retry crates I found seemed to
+    /// either have way too much overhead or be very sketchy in terms of quality.
+    ///
+    /// # Arguments
+    ///
+    /// * `addr` - Address (more specifically path, but left address for clarity) of the IPC socket (same as `send_command`).
+    /// * `cmd` - Command to be sent to the daemon (same as `send_command`).
+    /// * `retry_delay_list` - a list of retry delays. The length of the list is the retry count.
+    /// # Returns
+    ///
+    /// A Result containing a response string (same as `send_command`).
+    pub fn send_command_with_retries(
```

Comment:
Is there really a retry here? 

I would imagine, that if socket failed to open, or any other `write()` returned an error, in this local setup we can probably consider this to be fatal error. But maybe there are some scenarios, which I am forgetting to consider?

Reply:
The daemon takes time to daemonize and start listening for commands and if we send commands before it's started, it returns "connection refused" errors. I've figured these retry times by experimenting with the startup times of the daemon. I might have missed something which was causing the startup delay, but when I was trying to debug it by measuring OS time at various points, everything seemed to make sense.

Reply:
OOoh, I see this is used at startup, as when the process is still being started. But I am still slightly confused why this is needed, as the socket file creation itself should be a good indicator for daemon process readyness.